File: week05/2/Prob4.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        result = ListNode()
        # if list is null
        if list1 == None:
            return list2
        if list2 == None:
            return list1
        
        # check val
        if list1.val < list2.val:
            result = list1
            result.next = self.mergeTwoLists(list1.next, list2)
        else:
            result = list2
            result.next = self.mergeTwoLists(list1, list2.next)
        
        # return node
        return result

File: week05/2/Prob4.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeTwoLists(ListNode* list1, ListNode* list2) {
        ListNode* result = nullptr;
        //if list is null
        if(list1 == nullptr) return list2;
        if(list2 == nullptr) return list1;

        //compare val in two lists
        if(list1->val < list2->val){
            result = list1;
            result->next = mergeTwoLists(list1->next, list2);
        }else{
            result = list2;
            result->next = mergeTwoLists(list1, list2->next);
        }

        return result;
    }
};